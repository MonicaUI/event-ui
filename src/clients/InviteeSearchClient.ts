import { RequestBuilder, fetchAndRetryIfServerBusy } from '../utils/wrappedFetchAndRetryIfServerTooBusy';
import { SESSION_HEADER_EVENT_ID } from '@cvent/event-ui-networking';
import { ServiceError } from '@cvent/event-ui-networking';
import Logger from '@cvent/nucleus-logging';

const LOG = new Logger('InviteeSearchClient');

/**
 * The client to manage interaction with invitee search service
 */
export default class InviteeSearchClient {
  environment: $TSFixMe;
  inviteeSearchUrl;
  request;
  eventId;
  constructor(baseUrl: $TSFixMe, eventId: $TSFixMe, environment: $TSFixMe) {
    this.inviteeSearchUrl = baseUrl + 'invitee_search/v1';
    this.eventId = eventId;
    this.environment = environment;
    this.request = new RequestBuilder({ url: this.inviteeSearchUrl }).header(SESSION_HEADER_EVENT_ID, eventId);
  }

  /**
   * Update the invitee time zone preference
   */
  async updateInviteeTimeZonePreference(inviteeId: $TSFixMe, timeZonePreference: $TSFixMe): Promise<$TSFixMe> {
    LOG.debug('update Invitee time zone preference');
    const request = this.request
      .url(`${this.inviteeSearchUrl}/invitee/${inviteeId}/timezonePreference`)
      .query('environment', this.environment)
      .query('eventId', this.eventId)
      .put()
      .json({
        timeZonePreference
      })
      .build();
    const response = await fetchAndRetryIfServerBusy(request);
    if (!response.ok) {
      const error = await ServiceError.create('update invitee time zone preference failed', response, request);
      LOG.info('Error during updating Invitee zone preference ', request.url, error.responseStatus, error.responseBody);
      throw error;
    }
    return response.status;
  }
}
