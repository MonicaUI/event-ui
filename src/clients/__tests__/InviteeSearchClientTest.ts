import InviteeSearchClient from '../InviteeSearchClient';

jest.mock('@cvent/nucleus-networking', () => {
  const RequestBuilderModule = jest.requireActual<$TSFixMe>('@cvent/nucleus-networking');
  class MockRequestBuilder extends RequestBuilderModule.RequestBuilder {
    constructor(requestOptions = {}) {
      const newRequestOptions = {
        ...requestOptions,
        headers: {
          ...(requestOptions as $TSFixMe).headers,
          httplogpageloadid: 'httplogpageloadid',
          httplogrequestid: 'httplogrequestid'
        }
      };
      super(newRequestOptions);
    }
  }
  let mockResponse = {};
  return {
    ...RequestBuilderModule,
    RequestBuilder: MockRequestBuilder,
    setMockResponse(response) {
      mockResponse = response;
    },
    fetchAndRetryIfServerBusy: jest.fn(() => {
      return Promise.resolve({ ok: true, json: () => Promise.resolve(mockResponse) });
    })
  };
});

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { setMockResponse } = require('@cvent/nucleus-networking');

test('InviteeSearchClient.update timezone preference matches snapshot', async () => {
  const client = new InviteeSearchClient('https://example.com/reg/', 'eventId', 'dev');
  setMockResponse({}, {});
  const response = await client.updateInviteeTimeZonePreference('inviteeId', 0);
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const fetch = require('@cvent/nucleus-networking').fetchAndRetryIfServerBusy;
  expect(fetch).toHaveBeenCalled();
  expect(fetch.mock.calls[0]).toMatchSnapshot();
  expect(response).toMatchSnapshot();
});
