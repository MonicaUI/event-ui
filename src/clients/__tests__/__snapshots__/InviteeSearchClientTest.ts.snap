// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InviteeSearchClient.update timezone preference matches snapshot 1`] = `
Array [
  Request {
    "_bodyInit": "{\\"timeZonePreference\\":0}",
    "_bodyText": "{\\"timeZonePreference\\":0}",
    "credentials": "omit",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
        "cvent-event-id": "eventId",
        "httplogpageloadid": "httplogpageloadid",
        "httplogrequestid": "httplogrequestid",
      },
    },
    "method": "PUT",
    "mode": null,
    "referrer": null,
    "url": "https://example.com/reg/invitee_search/v1/invitee/inviteeId/timezonePreference?environment=dev&eventId=eventId",
  },
]
`;

exports[`InviteeSearchClient.update timezone preference matches snapshot 2`] = `undefined`;
